### PR TITLE
 Performance improvement from ImmersivePedia project.

### PR DESCRIPTION
**Pull request by douglasalipio**
Improving immersivepedia demo

Wip - Improvements to reduce the demo initial loading time:

- Loading
 dinossaurs mesh textures asynchronosly.

### CODE REVIEW

File: gvr-immersivepedia/gvrimmersivepedia/src/main/project.properties [hunk 1]
```diff
@@ -12,4 +12,4 @@
 
 # Project target.
 target=android-19
-android.library.reference.1=../../../../../GearVRf/GVRf/Framework/framework/src/main
+android.library.reference.1=../../../../../../vr-accessibility-gvrf-api/GearVRf/GVRf/Framework
```
Comment (thomasflynn) [inline gvr-immersivepedia/gvrimmersivepedia/src/main/project.properties]: this should be reverted

**Comment by liaxim** [comment]
Helps a little bit. I think though that onInit needs to be rewritten to be able to have the GVRF logo show up as soon as possible. Basically onInit needs to use the approach from this pull request to load textures and meshes and when everything the MenuScene needs is ready, dismiss the splash screen and switch to the menu scene.

**Comment by douglasalipio** [comment]
Hi @liaxim 
I changed the format texture .png to etc_2. I believe it has improved.

**Comment by thomasflynn** [comment]
can you move the etc2 files out of res/drawable-xxxhdpi and into res/raw?  AndroidStudio seems to have a fit them.  If they're moved to raw/, it seems fine with them.  Startup time seems to be reduced to ~15 seconds, good.  Though, I don't see the splash screen in the meantime while loading.

**Comment by douglasalipio** [comment]
@thomasflynn  @liaxim

**Comment by liaxim** [comment]
I still see the following pkm files under drawable-xxxhdpi:
```
blueskybox.pkm
environment_tex_diffuse.pkm
ground_tex_diffuse.pkm
skybox.pkm
```

**Comment by liaxim** [comment]
Looks like only ground_tex_diffuse is used. Please remove the rest if they are really not needed.

**Comment by liaxim** [comment]
Tried building with Eclipse. I am getting crashes in libc:
```
05-06 10:11:43.095  7764  7854 D GVRAndroidResource: Looking for compressed header
05-06 10:11:43.105  7764  7855 F libc    : Fatal signal 11 (SIGSEGV), code 2, fault addr 0x762d2000 in tid 7855 (gvrf-2)
```

How does it work for you? You might consider moving the pkms under the assets folder and use the slightly different API to get to them.

**Comment by douglasalipio** [comment]
Answering your questions:

1 - I believe the files in the folder raw would be more interesting for that being the files
on asserts I have to create a structure to retrieve the file names, as they can not access the files on asserts through class R.

2 - I am build project with Eclipse. It has worked well.

File: gvr-immersivepedia/gvrimmersivepedia/src/main/java/org/gearvrf/immersivepedia/dinosaur/DinosaurFactory.java [hunk 2]
```diff
@@ -88,53 +91,91 @@ private Dinosaur createDinosauros(String name, int dinoMeshId, int dinoTextureId
...
+        }, new GVRAndroidResource(gvrContext, dinoTextureId));
+
         return dino;
     }
 
     private FocusableSceneObject createDinosaurBase(int baseMeshId) {
-
         GVRMesh baseMesh = gvrContext.loadMesh(new GVRAndroidResource(gvrContext, baseMeshId), settings);
-        GVRTexture baseTexture = gvrContext.loadTexture(new GVRAndroidResource(gvrContext, R.drawable.base_tex_diffuse));
+        GVRTexture baseTexture = gvrContext.loadTexture(new GVRAndroidResource(gvrContext, R.raw.base_tex_diffuse));
```
Comment (liaxim) [inline gvr-immersivepedia/gvrimmersivepedia/src/main/java/org/gearvrf/immersivepedia/dinosaur/DinosaurFactory.java]: This uses a version of loadTexture that expects a bitmap image as an input. You should be using an overload that explicitly mentions that it tries to detect the texture type. The app definitely crashes on my S7.

**Comment by liaxim** [comment]
If it works for everybody but me we can merge. I will update it afterwards to get it to run on my S7.

**Comment by douglasalipio** [comment]
Hi @liaxim 
I made these modifications testing on my Note4. I will try to build on S7.

**Comment by fredduarte** [comment]
@liaxim 
We've tested with an S7 here and the demo built successfully (on Eclipse).
Could you give us a more detailed log on the problem?

**Comment by liaxim** [comment]
Here it is:

```
05-10 13:24:34.768  3221  3221 F DEBUG   : pid: 5734, tid: 6245, name: GLThread 7341  >>> com.samsung.android.gvr.sample                                                          <<<
05-10 13:24:34.768  3221  3221 F DEBUG   : signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0x762d3000
05-10 13:24:34.798  3221  3221 F DEBUG   :     r0 762d3000  r1 debc3744  r2 0000ffcc  r3 0000fff0
05-10 13:24:34.798  3221  3221 F DEBUG   :     r4 0000e44c  r5 02000000  r6 00000000  r7 d55b4e94
05-10 13:24:34.798  3221  3221 F DEBUG   :     r8 762d18dc  r9 762d18dc  sl 00000000  fp f722de07
05-10 13:24:34.798  3221  3221 F DEBUG   :     ip 20000000  sp da5fe1f8  lr f7224805  pc f727c850  cpsr 20070010
05-10 13:24:34.808  3221  3221 F DEBUG   : 
05-10 13:24:34.808  3221  3221 F DEBUG   : backtrace:
05-10 13:24:34.808  3221  3221 F DEBUG   :     #00 pc 00017850  /system/lib/libc.so (__memcpy_base_aligned+84)
05-10 13:24:34.808  3221  3221 F DEBUG   :     #01 pc 0001c801  /system/lib/libandroidfw.so (android::StreamingZipInflater::read(void*, unsigned int)+72)
05-10 13:24:34.808  3221  3221 F DEBUG   :     #02 pc 00010b05  /system/lib/libandroidfw.so (android::_CompressedAsset::read(void*, unsigned int)+14)
05-10 13:24:34.808  3221  3221 F DEBUG   :     #03 pc 0009e1b5  /system/lib/libandroid_runtime.so (android::AssetStreamAdaptor::read(void*, unsigned int)+104)
05-10 13:24:34.808  3221  3221 F DEBUG   :     #04 pc 001fe46d  /system/lib/libskia.so (SkETC1ImageDecoder::onDecode(SkStream*, SkBitmap*, SkImageDecoder::Mode)+336)
05-10 13:24:34.808  3221  3221 F DEBUG   :     #05 pc 0018313f  /system/lib/libskia.so (SkImageDecoder::decode(SkStream*, SkBitmap*, SkColorType, SkImageDecoder::Mode)+38)
05-10 13:24:34.808  3221  3221 F DEBUG   :     #06 pc 00094b5f  /system/lib/libandroid_runtime.so
05-10 13:24:34.808  3221  3221 F DEBUG   :     #07 pc 00095093  /system/lib/libandroid_runtime.so
05-10 13:24:34.808  3221  3221 F DEBUG   :     #08 pc 7434d79b  /data/dalvik-cache/arm/system@framework@boot.oat (offset 0x2f1b000)
```

And this:

```
05-10 13:24:32.868  5734  6245 V GVRViewManager: onSurfaceChanged
05-10 13:24:33.268  5734  6279 I OVR     : DeviceManagerThread - poll(fds,1,994) = 0 (Tid=6279)
05-10 13:24:33.278  5734  5734 I Vsync   : Vsync 487481: Choreographer: 10 x 16.94 ms (59.50 Hz, at 8262.390)
05-10 13:24:33.278  5734  5745 I art     : Background sticky concurrent mark sweep GC freed 21828(2MB) AllocSpace objects, 18(424KB) LOS objects, 0% free, 103MB/103MB, paused 485us total 243.339ms
05-10 13:24:33.288  5734  5734 I Vsync   : Vsync 487482: Choreographer:  1 x 15.30 ms (59.50 Hz, at 8262.400)
05-10 13:24:33.538  5734  6294 I VrApi   : FPS=0,Prd=0ms,Tear=0,Early=0,Stale=0,VSnc=1,Lat=0,CPU4/GPU=-1/-1,2288/0MHz,TA=0/0/0/0,Mem=1794MHz,Free=1319MB,PSM=0,PLS=0,Temp=28.3C/0.0C
05-10 13:24:33.788  5734  5734 I Vsync   : Vsync 487512: Choreographer:  2 x 16.79 ms (59.57 Hz, at 8262.904)
05-10 13:24:34.058  5734  6245 D SplashScreen: currentTime = 1,462,911,874,071,429,767, timeout = 1,462,911,879,071,429,767
05-10 13:24:34.078  5734  6385 V MediaPlayer: decode(28, 43600946, 142260)
05-10 13:24:34.248  5734  6385 V MediaPlayer: decode(56, 43743269, 140590)
05-10 13:24:34.268  5734  6279 I OVR     : DeviceManagerThread - poll(fds,1,1000) = 0 (Tid=6279)
05-10 13:24:34.338  5734  6385 V MediaPlayer: decode(57, 43883917, 123890)
05-10 13:24:34.408  5734  6385 V MediaPlayer: decode(58, 44007870, 40691)
05-10 13:24:34.488  5734  6385 V MediaPlayer: decode(59, 44048625, 37351)
05-10 13:24:34.508  5734  6245 I MediaPlayer: Need to enable context aware info
05-10 13:24:34.508  5734  6245 V MediaPlayer-JNI: native_setup
05-10 13:24:34.508  5734  6245 V MediaPlayer: constructor
05-10 13:24:34.508  5734  6245 V MediaPlayer: setListener
05-10 13:24:34.518  5734  6245 V MediaPlayer-JNI: setParameter: key 1400
05-10 13:24:34.518  5734  6245 V MediaPlayer: MediaPlayer::setParameter(1400)
05-10 13:24:34.518  5734  6245 V MediaPlayer-JNI: setAudioStreamType: 3
05-10 13:24:34.518  5734  6245 V MediaPlayer: MediaPlayer::setAudioStreamType
05-10 13:24:34.518  5734  6245 V MediaPlayer-JNI: set_session_id(): 30
05-10 13:24:34.518  5734  6245 V MediaPlayer: MediaPlayer::setAudioSessionId(30)
05-10 13:24:34.518  5734  6245 V MediaPlayer-JNI: setDataSourceFD: fd 54
05-10 13:24:34.518  5734  6245 V MediaPlayer: setDataSource(54, 36681551, 6634686)
05-10 13:24:34.518  3228  3630 V MediaPlayerService: Create new client(7) from pid 5734, uid 10276, 
05-10 13:24:34.528  5734  6245 V MediaPlayer: setVideoSurfaceTexture
05-10 13:24:34.528  5734  6245 V MediaPlayer: prepare
05-10 13:24:34.548  5734  6294 I VrApi   : FPS=0,Prd=0ms,Tear=0,Early=0,Stale=0,VSnc=1,Lat=0,CPU4/GPU=-1/-1,2288/0MHz,TA=0/0/0/0,Mem=1794MHz,Free=1304MB,PSM=0,PLS=0,Temp=28.3C/0.0C
05-10 13:24:34.558  5734  5947 V MediaPlayer: message received msg=200, ext1=10973, ext2=0
05-10 13:24:34.558  5734  5947 W MediaPlayer: info/warning (10973, 0)
05-10 13:24:34.558  5734  5947 V MediaPlayer: callback application
05-10 13:24:34.558  5734  5947 V MediaPlayer: back from callback
05-10 13:24:34.558  5734  5947 V MediaPlayer: message received msg=1, ext1=0, ext2=0
05-10 13:24:34.558  5734  5947 V MediaPlayer: prepared
05-10 13:24:34.558  5734  5947 V MediaPlayer: signal application thread
05-10 13:24:34.558  5734  5947 V MediaPlayer: callback application
05-10 13:24:34.558  5734  5947 V MediaPlayer: back from callback
05-10 13:24:34.558  5734  5734 D MediaPlayer: setSubtitleAnchor in MediaPlayer
05-10 13:24:34.558  5734  6245 V MediaPlayer: prepare complete - status=0
05-10 13:24:34.558  5734  6245 D MediaPlayer: setSubtitleAnchor in MediaPlayer
05-10 13:24:34.568  5734  6245 V MediaPlayer: invoke 68
05-10 13:24:34.568  5734  5734 V MediaPlayer: invoke 68
05-10 13:24:34.568  5734  6245 V MediaPlayer-JNI: setLooping: 1
05-10 13:24:34.568  5734  6245 V MediaPlayer: MediaPlayer::setLooping
05-10 13:24:34.578  5734  6245 V MediaPlayer-JNI: setVolume: left 1.000000  right 1.000000
05-10 13:24:34.578  5734  6245 V MediaPlayer: MediaPlayer::setVolume(1.000000, 1.000000)
05-10 13:24:34.578  5734  6245 V MediaPlayer-JNI: start
05-10 13:24:34.578  5734  6245 V MediaPlayer: start
05-10 13:24:34.578  5734  6093 V MediaPlayer: message received msg=300, ext1=0, ext2=0
05-10 13:24:34.578  5734  6093 V MediaPlayer: Received SEC_MM_PLAYER_CONTEXT_AWARE
05-10 13:24:34.578  5734  6093 V MediaPlayer: callback application
05-10 13:24:34.578  5734  6093 V MediaPlayer: back from callback
05-10 13:24:34.578  5734  5734 I MediaPlayer: send context aware event
05-10 13:24:34.588  5734  5734 I MediaPlayer: sendBroadcast CONTEXT_AWARE_MUSIC_INFO - type(start) - id (7)
05-10 13:24:34.588  5734  6385 V MediaPlayer: decode(60, 44086046, 40691)
05-10 13:24:34.678  5734  6245 D ResourceCache: put resource CLOSED{filePath=null; resourceId=7f040033; assetPath=null, url=null} to cache
05-10 13:24:34.678  5734  6245 D ResourceCache: put resource CLOSED{filePath=null; resourceId=7f020011; assetPath=null, url=null} to cache
05-10 13:24:34.688  5734  6453 D GVRAndroidResource: Looking for compressed header
05-10 13:24:34.698  5734  6245 D ResourceCache: put resource CLOSED{filePath=null; resourceId=7f040031; assetPath=null, url=null} to cache
05-10 13:24:34.698  5734  6245 F libc    : Fatal signal 11 (SIGSEGV), code 2, fault addr 0x762d3000 in tid 6245 (GLThread 7341)
05-10 13:24:34.708  5734  6456 F libc    : Fatal signal 11 (SIGSEGV), code 1, fault addr 0xccccccd8 in tid 6456 (gvrf-2)
```

Since this is only happening on my phone using a somewhat special image, I think we should not delay this pull request anymore. It does work on my release Note5 and it works for Tom. We can handle this issue separately if necessary.

**Comment by liaxim** [comment]
Before this can be merged though, you do need to revert the change to project.properties. It would be nice if some of the commented out code is cleaned up too. You may consider my note about the loadTexture overload used but this is totally optional. Thanks.

**Comment by thomasflynn** [comment]
let me re-test on my S7.  i don't remember which device i tested this on.

**Comment by thomasflynn** [comment]
i think i ran this by clicking on the icon on the phone before, but if i try to run it via Android Studio, i get this:
Error while executing: am start -n "com.samsung.android.gvr.sample/org.gearvrf.immersivepedia.MainActivity" -a android.intent.action.MAIN -c android.intent.category.LAUNCHER
Starting: Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] cmp=com.samsung.android.gvr.sample/org.gearvrf.immersivepedia.MainActivity }
Error type 3
Error: Activity class {com.samsung.android.gvr.sample/org.gearvrf.immersivepedia.MainActivity} does not exist.

Error while Launching activity

**Comment by thomasflynn** [comment]
and, yes, it crashes on my Adreno based S7 and my Mali based S7 as well.

**Comment by douglasalipio** [comment]
@thomasflynn @liaxim

**Comment by thomasflynn** [comment]
Run on my S6, but crashes on both S7s (Euro and US).  Stacktrace is:
05-16 19:13:03.315  3224  3224 F DEBUG   :     #00 pc 00017860  /system/lib/libc.so (__memcpy_base_aligned+100)
05-16 19:13:03.315  3224  3224 F DEBUG   :     #01 pc 00010b27  /system/lib/libandroidfw.so (android::_CompressedAsset::read(void*, unsigned int)+48)
05-16 19:13:03.315  3224  3224 F DEBUG   :     #02 pc 0009e1a1  /system/lib/libandroid_runtime.so (android::AssetStreamAdaptor::read(void*, unsigned int)+104)
05-16 19:13:03.315  3224  3224 F DEBUG   :     #03 pc 001fe46d  /system/lib/libskia.so (SkETC1ImageDecoder::onDecode(SkStream*, SkBitmap*, SkImageDecoder::Mode)+336)
05-16 19:13:03.315  3224  3224 F DEBUG   :     #04 pc 0018313f  /system/lib/libskia.so (SkImageDecoder::decode(SkStream*, SkBitmap*, SkColorType, SkImageDecoder::Mode)+38)
05-16 19:13:03.315  3224  3224 F DEBUG   :     #05 pc 00094b4b  /system/lib/libandroid_runtime.so
05-16 19:13:03.315  3224  3224 F DEBUG   :     #06 pc 0009507f  /system/lib/libandroid_runtime.so
05-16 19:13:03.315  3224  3224 F DEBUG   :     #07 pc 7431079b  /data/dalvik-cache/arm/system@framework@boot.oat (offset 0x2f1a000)

**Comment by thomasflynn** [comment]
ok, figured out the deal with the crashes on S7.  boils down to using gvrContext.loadTexture(AndroidResource) with a file that is compressed.  To load a compressed file, you must use gvrContext.loadTexture(new TextureCallback());  

Given the stacktrace, it seems like even though it falls back to a Bitmap.decode(), that Android actually handled loading an ETC1 in some (earlier?) versions.  On the S7 with Marshmallow, that path crashes.

Let's get this done.  For now, change the spots where you're using gvrContext.loadTexture() without a TextureCallback() to use a R.drawable png rather than a R.raw pkm file.

**Comment by thomasflynn** [comment]
i created a new pull request that has your changes plus the ones needed for S7: https://github.com/gearvrf/GearVRf-Demos/pull/158

**Pull request closed without merging**